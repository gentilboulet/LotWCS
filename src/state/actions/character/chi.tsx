import { createAction } from "typesafe-actions";
import { historyMetaCreator } from "../meta";
import { ICost } from "../../models/character/costs";
import { TChiName } from "../../../data/chi";

export const chiBuy = createAction(
  "chi/BUY",
  (chi: TChiName, value: number, cost: ICost) => {
    return { chi, value, cost };
  },
  historyMetaCreator,
)();
