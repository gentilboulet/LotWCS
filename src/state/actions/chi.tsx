import { ICost } from 'costs/types';
import { TChiName } from 'data/chi';
import * as constants from 'state/constants/chi';

export interface IChiBuy {
  type: constants.CHI_BUY;
  chi: TChiName;
  value: number;
  cost: ICost;
}

export type IChiAction = IChiBuy;

export function chiBuy(chi: TChiName, value:number, cost: ICost): IChiBuy {
  return { chi, cost, type: constants.CHI_BUY, value };
}
