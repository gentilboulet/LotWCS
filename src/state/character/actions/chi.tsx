import { createAction } from "typesafe-actions";
import { TChiName } from "../../data/chi";
import { ICost } from "../costs";

export const chiBuy = createAction(
  "chi/BUY",
  action => (chi: TChiName, value: number, cost: ICost) => {
    return action({ chi, value, cost });
  }
);
