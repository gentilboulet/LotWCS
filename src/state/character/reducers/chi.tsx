import { produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/chi";
import { TChiState } from "../models/chi";

import { increase } from "../models/chi";

export type IChiAction = ActionType<typeof actions>;

export function chiReducer(chi: TChiState, action: IChiAction): TChiState {
  return produce(chi, draft => {
    switch (action.type) {
      case getType(actions.chiBuy):
        increase(draft, action.payload.chi, action.payload.value);
        break;
      default:
    }
  });
}
