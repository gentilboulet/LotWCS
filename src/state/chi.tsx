import { chiNames, TChiName } from 'data/chi';

export type TChiState = {
  [chi in TChiName]: { value: number, cultivation: number};
}

export function createState(): TChiState {
  const state = {} as TChiState;
  chiNames.forEach((chi: TChiName) => {
    state[chi] = { value: 0, cultivation: 0};
  });
  return state;
}

export function increase(state: TChiState, name: TChiName, value: number): void {
   state[name].value += value;
}

export function increaseCultivation(state: TChiState, name: TChiName, value: number): void {
  const actualCultivation = state[name].cultivation;
  const actualChi = state[name].value;

  let newCultivation = actualCultivation + value;
  let newChiValue = actualChi;
  while (newCultivation >= newChiValue) {
    newChiValue ++;
    newCultivation -= newChiValue;
  }
  state[name].value = newChiValue;
  state[name].cultivation = newCultivation;
}
