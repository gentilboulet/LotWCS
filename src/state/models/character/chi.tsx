import { chiNames, TChiName } from "../../../data/chi";

export type TChiState = {
  [chi in TChiName]: { value: number; cultivation: number };
};

export function createState(): TChiState {
  const state = {} as TChiState;
  chiNames.forEach((chi: TChiName) => {
    state[chi] = { value: 0, cultivation: 0 };
  });
  return state;
}

export function increase(
  state: TChiState,
  name: TChiName,
  value: number,
): void {
  state[name].value += value;
}

export function increaseCultivation(
  state: TChiState,
  name: TChiName,
  increaseValue: number,
): void {
  let cultivation = state[name].cultivation + increaseValue;
  let chiValue = state[name].value;

  const costMultiplier =
    name === "enlightened" || name === "corrupt"
      ? 10
      : name !== "general"
      ? 5
      : 1;

  while (cultivation - costMultiplier * Math.max(1, chiValue) >= 0) {
    cultivation -= costMultiplier * Math.max(1, chiValue);
    chiValue++;
  }

  state[name].value = chiValue;
  state[name].cultivation = cultivation;
}

export function canBuyChi(): boolean {
  return true;
}
