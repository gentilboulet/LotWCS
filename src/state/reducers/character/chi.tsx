import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/chi";
import { TChiState } from "../../models/character/chi";

import { increase } from "../../models/character/chi";

export type IChiAction = ActionType<typeof actions>;

export const chiReducer = produce(
  (draft: Draft<TChiState>, action: IChiAction) => {
    switch (action.type) {
      case getType(actions.chiBuy):
        increase(draft, action.payload.chi, action.payload.value);
        break;
      default:
    }
  },
);
