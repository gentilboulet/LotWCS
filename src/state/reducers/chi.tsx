import { produce } from 'immer';

import { IChiAction } from 'state/actions/chi';
import { IStoreState } from 'state/type';

import * as constants from 'state/constants/chi';

import { increase } from 'state/chi';

export function chiReducer(baseState: IStoreState, action: IChiAction): IStoreState {
  switch (action.type) {
    case constants.CHI_BUY:
      return produce(baseState, draftState => {
        // applyCost(state, action.cost);
        increase(draftState.chi, action.chi, action.value);
        draftState.history.push(action);
      });
    default:
  }
  return baseState;
}
