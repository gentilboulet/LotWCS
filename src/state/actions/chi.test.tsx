import * as actions from './chi';

describe('Testing chi action creator', () => {
  const noCost = {
    destiny: 0, discountIdx: -1, discountNewValue: 0,
    entanglement: 0
  };

  it('should create a chi buy action', () => {
    const chiType = 'fire';
    const value = 123;
    const action = actions.chiBuy(chiType, value, noCost);
    expect( action.chi ).toBe(chiType);
    expect( action.value ).toBe(value);
    expect( action ).toMatchSnapshot();
  });
});
