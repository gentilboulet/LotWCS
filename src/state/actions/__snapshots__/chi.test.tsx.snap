// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing chi action creator should create a chi buy action 1`] = `
Object {
  "chi": "fire",
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "type": "CHI_BUY",
  "value": 123,
}
`;
